### PR TITLE
Contextual Bandit algorithms (WIP)

### PR DESCRIPTION
**Pull request by saurabh3949**
## Why are these changes needed?

This PR adds the following contextual bandit algorithms to RLlib contrib:
- LinUCB
- LinTS (Thompson sampling)

Since these algorithms work with a linear model, I have added a pytorch based Ridge linear regression module that has a closed form solution and hence doesn't need an SGD based optimizer.

The PR also contains some environments that generate data synthetically for both discrete and parametric action spaces. 

Please see the future plan and tasks for Rllib bandits here:
https://github.com/saurabh3949/ray/projects/1

## Related issue number

None

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

This work is in progress. The missing checks will be added soon.

### CODE REVIEW

File: rllib/contrib/bandits/agents/linear.py [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+    UCB = 2
+
+
+class BanditPolicyOverrides:
+    @override(Policy)
+    def _create_exploration(self, action_space, config):
+        exploration_config = config.get("exploration_config",
+                                        {"type": "StochasticSampling"})
+        if exploration_config[
+                "type"] == ImplicitExploration.ThompsonSampling.name:
```
Comment (ericl) [inline rllib/contrib/bandits/agents/linear.py]: @sven1977 is this the right way to create ThompsonSampling or it something that can be auto-created from_config?

File: rllib/contrib/bandits/agents/__init__.py [hunk 2]
```diff
@@ -0,0 +1,3 @@
+from ray.rllib.contrib.bandits.agents.linear import LinTSTrainer, LinUCBTrainer
+
+__all__ = ["LinTSTrainer", "LinUCBTrainer"]
```
Comment (ericl) [inline rllib/contrib/bandits/agents/__init__.py]: Can you add these to contrib/registry.py so that it can be run from the command line?

File: rllib/contrib/bandits/agents/linear.py (hunk 1 above)
Comment (sven1977) [reply rllib/contrib/bandits/agents/linear.py]: Yes, we should create it from the config. I guess the problem is that `from_config` only looks in the package in which the base Exploration class is located and thus would miss these new custom classes. However, you can always do:
```
exploration_config: {
"type": "ray.rllib.contrib.bandits.exploration. ImplicitExploration.ThompsonSampling"
}
```

**Review by ericl — COMMENTED** [review]
@saurabh3949 I pushed some updates for better integration with gym envs (report rewards, make regret optional, register the algos).

For example, now you can `./train.py --env=CartPole-v0 --run=contrib/LinUCB` though it probably won't learn anything.

I think the only major missing element now is documentation. Could you add a section to `doc/source/rllib-algorithms.rst` on the basics of LinUCB / LinTS and how to use them effectively?

File: rllib/contrib/bandits/agents/linear.py (hunk 1 above)
Comment (ericl) [reply rllib/contrib/bandits/agents/linear.py]: @sven1977 could you push the patch for this?

**Review by saurabh3949 — COMMENTED** [review]
Thanks a lot @ericl . I am working on:
- Documentation
- GLM bandit algorithm
- Docstrings

**Comment by ericl** [comment]
Sounds good. By the way, I noticed that LinUCB doesn't converge to 10.0 reward on the test case I added: https://github.com/ray-project/ray/pull/7642/files#diff-a53cc7bd3d9294e8bee726da66d8b940R41

Is this expected?

File: rllib/contrib/bandits/agents/linear.py (hunk 1 above)
Comment (saurabh3949) [reply rllib/contrib/bandits/agents/linear.py]: I just pushed the patch for this,

**Review by saurabh3949 — COMMENTED** [review]
@ericl : The default UCB arguments work well for small reward range. To handle larger rewards, one would need to modify the hyperparams of LinUCB. I have rescaled the reward to be in [0,1] in the simple env.

**Comment by ericl** [comment]
Got it. Though it seems a reward range in [0, 1] now means LinTS doesn't converge to max. Would be great to make them consistent and recommend envs scale their rewards to a specific range.

File: rllib/contrib/bandits/agents/linear.py [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+
+def init_cum_regret(policy, *args):
+    policy.regrets = []
+
+
+# Build a common policy for TS and UCB
+LinUCBTSPolicy = build_torch_policy(
+    name="LinUCBTSPolicy",
+    get_default_config=lambda: DEFAULT_CONFIG,
+    loss_fn=None,
```
Comment (sven1977) [inline rllib/contrib/bandits/agents/linear.py]: Should we add a `validate_config=validate_config_fn` here?
Or would it make sense to run these with explorations different from UCB/TS?

Comment (saurabh3949) [reply rllib/contrib/bandits/agents/linear.py]: It is possible to use this policy with any exploration config. It does make sense as one might want to compare e-greedy or softmax sampling with UCB or TS.

**Review by sven1977 — APPROVED** [review]
This looks really great already! Let's get this merged. We can add actual regression tests in a follow-up PR.

**Comment by ericl** [comment]
Sounds good, merging.

**Comment by ericl** [comment]
Note that we do have two regression tests added on the SimpleContextBandit env -- I think the main missing piece for this to be usable is documentation on how to configure the bandit algos for best performance, suitable envs, etc.

**Comment by sven1977** [comment]
Wait, we have examples, but nothing in `tuned_examples/regression`, correct?

**Comment by ericl** [comment]
Yep, but I added these to the BUILD: https://github.com/ray-project/ray/pull/7642/files#diff-ec6b944926873ea348809d63ef2fbd00R1371

The reason they aren't regression tests is since it requires a custom environment. Potentially we could also solve say, CartPole, with bandits, though I haven't looked into if it's possible.

**Comment by saurabh3949** [comment]
Thanks for merging this. I am working on adding documentation on how to configure the bandit algos for best performance, suitable envs, etc.
I think it will be hard or nearly impossible for bandits to solve CartPole, as bandits assume that the actions don't affect the state of the env.
